dune/iga/nurbstrimfunctionality.hh

```diff
@@ -0,0 +1,303 @@
+//
+// Created by Henri on 14.03.2023.
+//
+
+#pragma once
+
+namespace Dune::IGA::Impl::Trim {
+
+  using ctype        = double;
+  using ClipperPoint = Clipper2Lib::Point<ctype>;
+  using Point        = FieldVector<ctype, 2>;
+  using PointVector  = std::vector<Point>;
+
+  // To keep track of which point belongs to which edge or node we will use a pointMap
+  struct IntersectionPoint {
+    Point point;
+    bool alreadyVisited = false;
+
+    explicit IntersectionPoint(ClipperPoint& _clipperPoint) : point({_clipperPoint.x, _clipperPoint.y}){};
+  };
+
+  using IntersectionPointMap    = std::map<int, std::vector<IntersectionPoint>>;
+  using IntersectionPointMapPtr = std::shared_ptr<std::map<int, std::vector<IntersectionPoint>>>;
+
+  PointVector getElementCorners(const auto& element) {
+    auto geo = element.geometry();
+
+    std::vector<Point> corners(4);
+    corners[0] = geo.corner(0);
+    corners[1] = geo.corner(1);
+    corners[2] = geo.corner(3);  // see dune book page 127 Figure 5.12
+    corners[3] = geo.corner(2);
+
+    return corners;
+  }
+
+  Clipper2Lib::PathD getElementEdges(const auto& element) {
+    auto corners = getElementCorners(element);
+    Clipper2Lib::PathD edges;
+    for (int i = 0; i < 4; ++i)
+      edges.emplace_back(corners[i][0], corners[i][1]);
+    return edges;
+  }
+
+  bool areCurvesOnTheSameLine(NurbsCurveHandler& curve1, NurbsCurveHandler& curve2) {
+    const double tolerance = double(16) * std::numeric_limits<double>::epsilon();
+
+    auto domain1 = curve1.domain();
+    auto domain2 = curve2.domain();
+
+    auto p1 = curve1(domain1[0]);
+    auto p2 = curve1(domain1[1]);
+
+    auto p3 = curve2(domain2[0]);
+    auto p4 = curve2(domain2[1]);
+
+    // Compute direction vectors of both curves
+    auto d1 = p2 - p1;
+    auto d2 = p4 - p3;
+
+    auto cos_angle = Dune::dot(d1, d2) / (d1.two_norm() * d2.two_norm());
+    auto angle     = std::acos(std::min(std::max(cos_angle, -1.0), 1.0));
+
+    // Check if the angle is small enough to consider the curves on the same line
+    return ((angle < tolerance) || (std::fabs(angle - std::numbers::pi) < tolerance));
+  }
+
+  struct IntersectionResult {
+    enum ResultFlag { Found, NotFound };
```

Comment:
Do you have more then these two cases?  If not you can replace maybe with a `bool found`?